File: grist_integration.js
/**
 * Integration script to send scraper output to Grist.
 * 
 * This script reads JSON/CSV files from scrapers and sends them to Grist tables.
 * It can be used as a wrapper around existing scrapers or called after scrapers run.
 */

const fs = require('fs').promises;
const path = require('path');
const { GristClient, inferColumnsFromData } = require('./grist_client');

/**
 * Load data from JSON file.
 */
async function loadJsonData(filePath) {
    const content = await fs.readFile(filePath, 'utf-8');
    const data = JSON.parse(content);
    
    // Handle both list and dict formats
    if (Array.isArray(data)) {
        return data;
    } else if (typeof data === 'object') {
        // If it's a dict, try to find a list value
        for (const value of Object.values(data)) {
            if (Array.isArray(value)) {
                return value;
            }
        }
        // If no list found, wrap the dict
        return [data];
    }
    
    return [data];
}

/**
 * Load data from CSV file.
 */
async function loadCsvData(filePath) {
    // For CSV, we'll use a simple parser or require a library
    // For now, let's use a basic approach
    const content = await fs.readFile(filePath, 'utf-8');
    const lines = content.trim().split('\n');
    if (lines.length === 0) return [];
    
    const headers = lines[0].split(',').map(h => h.trim());
    const data = [];
    
    for (let i = 1; i < lines.length; i++) {
        const values = lines[i].split(',').map(v => v.trim());
        const record = {};
        headers.forEach((header, index) => {
            record[header] = values[index] || null;
        });
        data.push(record);
    }
    
    return data;
}

/**
 * Send data to Grist table.
 * 
 * @param {Array<Object>} data - List of record objects
 * @param {string} apiKey - Grist API key
 * @param {string} server - Grist server URL
 * @param {string} docName - Document name
 * @param {string} tableName - Table name
 * @param {string} org - Organization name
 * @param {boolean} upsert - Whether to upsert (update existing) records
 * @param {Array<string>} keyColumns - Columns to use for upsert matching
 */
async function sendToGrist(data, apiKey, server, docName, tableName, org = 'brightstar', upsert = false, keyColumns = null) {
    if (!data || data.length === 0) {
        console.log('No data to send');
        return;
    }

    const client = new GristClient(apiKey, server, org);

    // Get or create document
    console.log(`Getting/creating document: ${docName}`);
    const doc = await client.getOrCreateDocument(docName);
    console.log(`Document ID: ${doc.id}`);

    // Infer columns from data
    console.log('Inferring column types from data...');
    const columns = inferColumnsFromData(data);
    console.log(`Detected ${columns.length} columns: ${columns.map(c => c.id).join(', ')}`);

    // Ensure table exists with correct schema
    console.log(`Ensuring table '${tableName}' exists...`);
    await client.ensureTable(doc.id, tableName, columns);

    // Add timestamp column if not present
    const hasTimestamp = columns.some(c => c.id === 'scraped_at');
    if (!hasTimestamp) {
        await client.ensureTable(doc.id, tableName, [{ id: 'scraped_at', type: 'DateTime' }]);
    }

    // Add timestamp to all records
    const timestamp = new Date().toISOString();
    data.forEach(record => {
        if (!record.scraped_at) {
            record.scraped_at = timestamp;
        }
    });

    // Send data
    console.log(`Sending ${data.length} records to Grist...`);
    let result;
    if (upsert) {
        result = await client.upsertRecords(doc.id, tableName, data, keyColumns);
    } else {
        result = await client.addRecords(doc.id, tableName, data);
    }

    console.log(`✅ Successfully sent ${data.length} records to ${docName}/${tableName}`);
    return result;
}

/**
 * Main function - CLI interface
 */
async function main() {
    const args = process.argv.slice(2);
    
    if (args.length === 0 || args.includes('--help') || args.includes('-h')) {
        console.log(`
Usage: node grist_integration.js <input_file> [options]

Options:
  --api-key <key>        Grist API key (required)
  --server <url>         Grist server URL (required)
  --doc <name>           Grist document name (default: ABS_Scraper_Data)
  --table <name>         Grist table name (required)
  --org <name>           Grist organization name (default: brightstar)
  --upsert               Upsert records (update existing)
  --key-columns <cols>   Columns to use for upsert matching (space-separated)

Example:
  node grist_integration.js msm_results.json \\
    --api-key YOUR_KEY \\
    --server https://grist.pythonfinancial.com \\
    --doc "ABS_Data" \\
    --table "MSM_Results"
        `);
        process.exit(0);
    }

    // Parse arguments
    const inputFile = args[0];
    let apiKey = null;
    let server = null;
    let docName = 'ABS_Scraper_Data';
    let tableName = null;
    let org = 'brightstar';
    let upsert = false;
    let keyColumns = null;

    for (let i = 1; i < args.length; i++) {
        if (args[i] === '--api-key' && args[i + 1]) {
            apiKey = args[++i];
        } else if (args[i] === '--server' && args[i + 1]) {
            server = args[++i];
        } else if (args[i] === '--doc' && args[i + 1]) {
            docName = args[++i];
        } else if (args[i] === '--table' && args[i + 1]) {
            tableName = args[++i];
        } else if (args[i] === '--org' && args[i + 1]) {
            org = args[++i];
        } else if (args[i] === '--upsert') {
            upsert = true;
        } else if (args[i] === '--key-columns' && args[i + 1]) {
            keyColumns = args[++i].split(/\s+/);
        }
    }

    // Validate required arguments
    if (!apiKey) {
        console.error('Error: --api-key is required');
        process.exit(1);
    }
    if (!server) {
        console.error('Error: --server is required');
        process.exit(1);
    }
    if (!tableName) {
        console.error('Error: --table is required');
        process.exit(1);
    }

    // Load data
    const inputPath = path.resolve(inputFile);
    try {
        await fs.access(inputPath);
    } catch (e) {
        console.error(`Error: File not found: ${inputFile}`);
        process.exit(1);
    }

    console.log(`Loading data from ${inputFile}...`);
    let data;
    const ext = path.extname(inputFile).toLowerCase();
    
    if (ext === '.json') {
        data = await loadJsonData(inputPath);
    } else if (ext === '.csv') {
        data = await loadCsvData(inputPath);
    } else {
        console.error(`Error: Unsupported file format: ${ext}`);
        console.error('Supported formats: .json, .csv');
        process.exit(1);
    }

    console.log(`Loaded ${data.length} records`);

    // Send to Grist
    try {
        await sendToGrist(data, apiKey, server, docName, tableName, org, upsert, keyColumns);
    } catch (error) {
        console.error(`Error sending data to Grist: ${error.message}`);
        if (error.stack) {
            console.error(error.stack);
        }
        process.exit(1);
    }
}

// Run if called directly
if (require.main === module) {
    main().catch(error => {
        console.error('Unhandled error:', error);
        process.exit(1);
    });
}

module.exports = { sendToGrist, loadJsonData, loadCsvData };

File: grist_integration.py
"""
Integration script to send scraper output to Grist.

This script reads JSON/CSV files from scrapers and sends them to Grist tables.
It can be used as a wrapper around existing scrapers or called after scrapers run.
"""

import os
import sys
import json
import csv
import argparse
from pathlib import Path
from grist_client import GristClient, infer_columns_from_data


def load_json_data(file_path: str) -> list:
    """Load data from JSON file."""
    with open(file_path, 'r', encoding='utf-8') as f:
        data = json.load(f)
        # Handle both list and dict formats
        if isinstance(data, dict):
            # If it's a dict, try to find a list value
            for key, value in data.items():
                if isinstance(value, list):
                    return value
            # If no list found, wrap the dict
            return [data]
        return data if isinstance(data, list) else [data]


def load_csv_data(file_path: str) -> list:
    """Load data from CSV file."""
    data = []
    with open(file_path, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        for row in reader:
            # Convert empty strings to None for better type inference
            cleaned_row = {k: (v if v else None) for k, v in row.items()}
            data.append(cleaned_row)
    return data


def send_to_grist(
    data: list,
    api_key: str,
    server: str,
    doc_name: str,
    table_name: str,
    org: str = "brightstar",
    upsert: bool = False,
    key_columns: list = None
):
    """
    Send data to Grist table.
    
    Args:
        data: List of record dicts
        api_key: Grist API key
        server: Grist server URL
        doc_name: Document name
        table_name: Table name
        org: Organization name
        upsert: Whether to upsert (update existing) records
        key_columns: Columns to use for upsert matching
    """
    if not data:
        print("No data to send")
        return
    
    client = GristClient(api_key, server, org)
    
    # Get or create document
    print(f"Getting/creating document: {doc_name}")
    doc = client.get_or_create_document(doc_name)
    print(f"Document ID: {doc['id']}")
    
    # Infer columns from data
    print("Inferring column types from data...")
    columns = infer_columns_from_data(data)
    print(f"Detected {len(columns)} columns: {', '.join([c['id'] for c in columns])}")
    
    # Ensure table exists with correct schema
    print(f"Ensuring table '{table_name}' exists...")
    client.ensure_table(doc["id"], table_name, columns)
    
    # Add timestamp column if not present
    from datetime import datetime
    if "scraped_at" not in [c["id"] for c in columns]:
        client.ensure_table(doc["id"], table_name, [{"id": "scraped_at", "type": "DateTime"}])
    
    # Add timestamp to all records
    timestamp = datetime.now().isoformat()
    for record in data:
        if "scraped_at" not in record:
            record["scraped_at"] = timestamp
    
    # Send data
    print(f"Sending {len(data)} records to Grist...")
    if upsert:
        result = client.upsert_records(doc["id"], table_name, data, key_columns)
    else:
        result = client.add_records(doc["id"], table_name, data)
    
    print(f"✅ Successfully sent {len(data)} records to {doc_name}/{table_name}")
    return result


def main():
    parser = argparse.ArgumentParser(
        description="Send scraper output to Grist tables"
    )
    parser.add_argument("input_file", help="Input JSON or CSV file")
    parser.add_argument("--api-key", required=True, help="Grist API key")
    parser.add_argument("--server", required=True, help="Grist server URL")
    parser.add_argument("--doc", default="ABS_Scraper_Data", help="Grist document name")
    parser.add_argument("--table", required=True, help="Grist table name")
    parser.add_argument("--org", default="brightstar", help="Grist organization name")
    parser.add_argument("--upsert", action="store_true", help="Upsert records (update existing)")
    parser.add_argument("--key-columns", nargs="+", help="Columns to use for upsert matching")
    
    args = parser.parse_args()
    
    # Load data
    input_path = Path(args.input_file)
    if not input_path.exists():
        print(f"Error: File not found: {args.input_file}")
        sys.exit(1)
    
    print(f"Loading data from {args.input_file}...")
    if input_path.suffix.lower() == '.json':
        data = load_json_data(str(input_path))
    elif input_path.suffix.lower() == '.csv':
        data = load_csv_data(str(input_path))
    else:
        print(f"Error: Unsupported file format: {input_path.suffix}")
        print("Supported formats: .json, .csv")
        sys.exit(1)
    
    print(f"Loaded {len(data)} records")
    
    # Send to Grist
    try:
        send_to_grist(
            data,
            args.api_key,
            args.server,
            args.doc,
            args.table,
            args.org,
            args.upsert,
            args.key_columns
        )
    except Exception as e:
        print(f"Error sending data to Grist: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == "__main__":
    main()

